README.md
# e-ngx-echarts

基于 Angular 的 [echarts](http://echarts.baidu.com/index.html) 图表组件

## Usage

1. Install

	```shell
	npm install --save e-ngx-echarts
	```

2. Add the ENgxChartModule

	```typescript
	import { ENgxEChartsModule } from 'e-ngx-echarts';
	@NgModule({
	    imports: [
	        ENgxEChartsModule
	    ]
	})
	```

3. template

	```html
	<div style="width: 600px; height:400px;" echarts [options]="options" (ready)="onEChartsReady($event)"></div>
	```

4. component

	```typescript
	options = {
        title: {
            text: '堆叠区域图'
        },
        tooltip: {
            trigger: 'axis',
            axisPointer: {
                type: 'cross',
                label: {
                    backgroundColor: '#6a7985'
                }
            }
        },
        legend: {
            data: ['邮件营销', '联盟广告', '视频广告', '直接访问', '搜索引擎']
        },
        toolbox: {
            feature: {
                saveAsImage: {}
            }
        },
        grid: {
            left: '3%',
            right: '4%',
            bottom: '3%',
            containLabel: true
        },
        xAxis: [
            {
                type: 'category',
                boundaryGap: false,
                data: ['周一', '周二', '周三', '周四', '周五', '周六', '周日']
            }
        ],
        yAxis: [
            {
                type: 'value'
            }
        ],
        series: [
            {
                name: '邮件营销',
                type: 'line',
                stack: '总量',
                areaStyle: {normal: {}},
                data: [120, 132, 101, 134, 90, 230, 210]
            },
            {
                name: '联盟广告',
                type: 'line',
                stack: '总量',
                areaStyle: {normal: {}},
                data: [220, 182, 191, 234, 290, 330, 310]
            },
            {
                name: '视频广告',
                type: 'line',
                stack: '总量',
                areaStyle: {normal: {}},
                data: [150, 232, 201, 154, 190, 330, 410]
            },
            {
                name: '直接访问',
                type: 'line',
                stack: '总量',
                areaStyle: {normal: {}},
                data: [320, 332, 301, 334, 390, 330, 320]
            },
            {
                name: '搜索引擎',
                type: 'line',
                stack: '总量',
                label: {
                    normal: {
                        show: true,
                        position: 'top'
                    }
                },
                areaStyle: {normal: {}},
                data: [820, 932, 901, 934, 1290, 1330, 1320]
            }
        ]
    };
    
    onEChartsReady(evt: any) {
        console.log(evt);
    }
	```

## API

### exportAs

- `eNgxECharts` - 导出的指令变量，可在模板获取指令类并调用（`#eNgxECharts="eNgxECharts"`）。

### Inputs

- `options`（`Object`） - options 与[echarts的配置属性](http://echarts.baidu.com/option.html)一致

- `theme`（`Object`） - 调用 [echarts.init](http://echarts.baidu.com/api.html#echarts.init) 方法传入的主题的配置对象，也可以是使用已经通过 echarts.registerTheme 注册的主题名称。可通过[echarts主题构建工具](http://echarts.baidu.com/theme-builder/)导出自定义主题的配置对象（JSON版）

- `initOpts`（`Object`） - 调用 [echarts.init](http://echarts.baidu.com/api.html#echarts.init) 方法传入的附加参数

### Outputs

- `ready` - 初始化完成的事件，$event为当前 ENgxEChartsDirective 实例

- `chartClick`: 点击事件。参数同 `click` 事件

- `chartDblClick`: 双击事件。参数同 `dblclick` 事件

- `chartMouseDown`: 鼠标按下事件。`mousedown` 事件

- `chartMouseUp`: 鼠标松开事件。参数同 `mouseup` 事件

- `chartMouseOver`: 鼠标移上事件。参数同 `mouseover` 事件

- `chartMouseOut`: 鼠标移开事件。参数同 `mouseout` 事件

- `chartDataZoom`: 数据区域缩放后的事件。参数同 `dataZoom` 事件

更多事件可查看：[http://echarts.baidu.com/api.html#events](http://echarts.baidu.com/api.html#events)

### Properties

- `echartsIntance` - echarts 实例

### Service

- `ENgxEchartsService` -  全局 echarts 对象包装器，可以通过它获取全局 echarts 对象及相应方法

示例：

```typescript
import { ENgxEchartsService } from 'e-ngx-echarts';

//...
constructor(public es: ENgxEchartsService) {}

ngOnInit() {
	// 获取全局 echarts 对象 
	const echarts = this.es.echarts;
	echarts.registerTheme('wonderland', obj);
	
	// 或者直接调用服务包装好的方法：
	// this.es.registerTheme('wonderland', obj);
}
```

# License

[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](/LICENSE)


CHANGELOG.md
# 5.0.1（2018-07-26）

- 导出 ENgxEChartsDirective

# 5.0.0（2018-07-26）

- 发布正式版本


src/e-ngx-echarts.service.ts
import { Injectable } from '@angular/core';
import * as echarts from 'echarts';

@Injectable()
export class ENgxEchartsService {

	constructor() {
	}

	/**
	 * 获取全局 echarts 对象
	 */
	get echarts(): any {
		return echarts;
	}

	/**
	 * 包装 echarts.graphic
	 */
	get graphic(): any {
		return this._checkEcharts() ? echarts.graphic : undefined;
	}

	/**
	 * 包装 echarts.init
	 */
	get init(): any {
		return this._checkEcharts() ? echarts.init : undefined;
	}

	/**
	 * 包装 echarts.connect
	 */
	get connect(): any {
		return this._checkEcharts() ? echarts.connect : undefined;
	}

	/**
	 * 包装 echarts.disconnect
	 */
	get disconnect(): any {
		return this._checkEcharts() ? echarts.disconnect : undefined;
	}

	/**
	 * 包装 echarts.dispose
	 */
	get dispose(): any {
		return this._checkEcharts() ? echarts.dispose : undefined;
	}

	/**
	 * 包装 echarts.getInstanceByDom
	 */
	get getInstanceByDom(): any {
		return this._checkEcharts() ? echarts.getInstanceByDom : undefined;
	}

	/**
	 * 包装 echarts.registerMap
	 */
	get registerMap(): any {
		return this._checkEcharts() ? echarts.registerMap : undefined;
	}

	/**
	 * 包装 echarts.getMap
	 */
	get getMap(): any {
		return this._checkEcharts() ? echarts.getMap : undefined;
	}

	/**
	 * 包装 echarts.registerTheme
	 */
	get registerTheme(): any {
		return this._checkEcharts() ? echarts.registerTheme : undefined;
	}

	private _checkEcharts() {
		if (echarts) {
			return true;
		} else {
			console.error('[e-ngx-echarts] 全局 echarts 对象未加载！');
			return false;
		}
	}
}


src/e-ngx-echarts.directive.ts
import { Directive, ElementRef, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { ENgxEchartsService } from './e-ngx-echarts.service';

@Directive({
	selector: '[echarts]',
	exportAs: 'eNgxECharts'
})
export class ENgxEChartsDirective implements OnInit, OnDestroy {
	echartsIntance: any;

	@Input() options: any = {};
	@Input() theme: any;
	@Input() initOpts: any;
	@Output() ready: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartClick: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartDblClick: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartMouseDown: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartMouseMove: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartMouseUp: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartMouseOver: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartMouseOut: EventEmitter<any> = new EventEmitter<any>();
	@Output() chartDataZoom: EventEmitter<any> = new EventEmitter<any>();

	constructor(public el: ElementRef, public es: ENgxEchartsService) {
	}

	ngOnInit() {
		this.echartsIntance = this.es.init(this.el.nativeElement, this.theme, this.initOpts);
		this.echartsIntance.setOption(this.options);
		this.initEvents();
		this.ready.emit(this);
	}

	ngOnDestroy() {
		this.es.dispose(this.el.nativeElement);
	}

	initEvents() {
		this.echartsIntance.on('click', (params: any) => {
			this.chartClick.emit(params);
		});
		this.echartsIntance.on('dblclick', (params: any) => {
			this.chartDblClick.emit(params);
		});
		this.echartsIntance.on('mousedown', (params: any) => {
			this.chartMouseDown.emit(params);
		});
		this.echartsIntance.on('mousemove', (params: any) => {
			this.chartMouseMove.emit(params);
		});
		this.echartsIntance.on('mouseup', (params: any) => {
			this.chartMouseUp.emit(params);
		});
		this.echartsIntance.on('mouseover', (params: any) => {
			this.chartMouseOver.emit(params);
		});
		this.echartsIntance.on('mouseout', (params: any) => {
			this.chartMouseOut.emit(params);
		});
		this.echartsIntance.on('dataZoom', (params: any) => {
			this.chartDataZoom.emit(params);
		});
	}
}


example/app/app.component.ts
import { Component } from '@angular/core';
import { ENgxEchartsService } from '../../src';

@Component({
	selector: 'app-root',
	templateUrl: './app.component.html',
	styleUrls: ['./app.component.scss']
})
export class AppComponent {
	options = {
		title: {
			text: '堆叠区域图'
		},
		tooltip: {
			trigger: 'axis',
			axisPointer: {
				type: 'cross',
				label: {
					backgroundColor: '#6a7985'
				}
			}
		},
		legend: {
			data: ['邮件营销', '联盟广告', '视频广告', '直接访问', '搜索引擎']
		},
		toolbox: {
			feature: {
				saveAsImage: {}
			}
		},
		grid: {
			left: '3%',
			right: '4%',
			bottom: '3%',
			containLabel: true
		},
		xAxis: [
			{
				type: 'category',
				boundaryGap: false,
				data: ['周一', '周二', '周三', '周四', '周五', '周六', '周日']
			}
		],
		yAxis: [
			{
				type: 'value'
			}
		],
		series: [
			{
				name: '邮件营销',
				type: 'line',
				stack: '总量',
				areaStyle: {normal: {}},
				data: [120, 132, 101, 134, 90, 230, 210]
			},
			{
				name: '联盟广告',
				type: 'line',
				stack: '总量',
				areaStyle: {normal: {}},
				data: [220, 182, 191, 234, 290, 330, 310]
			},
			{
				name: '视频广告',
				type: 'line',
				stack: '总量',
				areaStyle: {normal: {}},
				data: [150, 232, 201, 154, 190, 330, 410]
			},
			{
				name: '直接访问',
				type: 'line',
				stack: '总量',
				areaStyle: {normal: {}},
				data: [320, 332, 301, 334, 390, 330, 320]
			},
			{
				name: '搜索引擎',
				type: 'line',
				stack: '总量',
				label: {
					normal: {
						show: true,
						position: 'top'
					}
				},
				areaStyle: {normal: {}},
				data: [820, 932, 901, 934, 1290, 1330, 1320]
			}
		]
	};

	constructor(public es: ENgxEchartsService) {
		es.echarts.registerTheme('wonderland', {
				'color': [
					'#4ea397',
					'#22c3aa',
					'#7bd9a5',
					'#d0648a',
					'#f58db2',
					'#f2b3c9'
				],
				'backgroundColor': 'rgba(255,255,255,0)',
				'textStyle': {},
				'title': {
					'textStyle': {
						'color': '#666666'
					},
					'subtextStyle': {
						'color': '#999999'
					}
				},
				'line': {
					'itemStyle': {
						'normal': {
							'borderWidth': '2'
						}
					},
					'lineStyle': {
						'normal': {
							'width': '3'
						}
					},
					'symbolSize': '8',
					'symbol': 'emptyCircle',
					'smooth': false
				},
				'radar': {
					'itemStyle': {
						'normal': {
							'borderWidth': '2'
						}
					},
					'lineStyle': {
						'normal': {
							'width': '3'
						}
					},
					'symbolSize': '8',
					'symbol': 'emptyCircle',
					'smooth': false
				},
				'bar': {
					'itemStyle': {
						'normal': {
							'barBorderWidth': 0,
							'barBorderColor': '#ccc'
						},
						'emphasis': {
							'barBorderWidth': 0,
							'barBorderColor': '#ccc'
						}
					}
				},
				'pie': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'scatter': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'boxplot': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'parallel': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'sankey': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'funnel': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'gauge': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						},
						'emphasis': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					}
				},
				'candlestick': {
					'itemStyle': {
						'normal': {
							'color': '#d0648a',
							'color0': 'transparent',
							'borderColor': '#d0648a',
							'borderColor0': '#22c3aa',
							'borderWidth': '1'
						}
					}
				},
				'graph': {
					'itemStyle': {
						'normal': {
							'borderWidth': 0,
							'borderColor': '#ccc'
						}
					},
					'lineStyle': {
						'normal': {
							'width': '1',
							'color': '#cccccc'
						}
					},
					'symbolSize': '8',
					'symbol': 'emptyCircle',
					'smooth': false,
					'color': [
						'#4ea397',
						'#22c3aa',
						'#7bd9a5',
						'#d0648a',
						'#f58db2',
						'#f2b3c9'
					],
					'label': {
						'normal': {
							'textStyle': {
								'color': '#ffffff'
							}
						}
					}
				},
				'map': {
					'itemStyle': {
						'normal': {
							'areaColor': '#eeeeee',
							'borderColor': '#999999',
							'borderWidth': 0.5
						},
						'emphasis': {
							'areaColor': 'rgba(34,195,170,0.25)',
							'borderColor': '#22c3aa',
							'borderWidth': 1
						}
					},
					'label': {
						'normal': {
							'textStyle': {
								'color': '#28544e'
							}
						},
						'emphasis': {
							'textStyle': {
								'color': 'rgb(52,158,142)'
							}
						}
					}
				},
				'geo': {
					'itemStyle': {
						'normal': {
							'areaColor': '#eeeeee',
							'borderColor': '#999999',
							'borderWidth': 0.5
						},
						'emphasis': {
							'areaColor': 'rgba(34,195,170,0.25)',
							'borderColor': '#22c3aa',
							'borderWidth': 1
						}
					},
					'label': {
						'normal': {
							'textStyle': {
								'color': '#28544e'
							}
						},
						'emphasis': {
							'textStyle': {
								'color': 'rgb(52,158,142)'
							}
						}
					}
				},
				'categoryAxis': {
					'axisLine': {
						'show': true,
						'lineStyle': {
							'color': '#cccccc'
						}
					},
					'axisTick': {
						'show': false,
						'lineStyle': {
							'color': '#333'
						}
					},
					'axisLabel': {
						'show': true,
						'textStyle': {
							'color': '#999999'
						}
					},
					'splitLine': {
						'show': true,
						'lineStyle': {
							'color': [
								'#eeeeee'
							]
						}
					},
					'splitArea': {
						'show': false,
						'areaStyle': {
							'color': [
								'rgba(250,250,250,0.05)',
								'rgba(200,200,200,0.02)'
							]
						}
					}
				},
				'valueAxis': {
					'axisLine': {
						'show': true,
						'lineStyle': {
							'color': '#cccccc'
						}
					},
					'axisTick': {
						'show': false,
						'lineStyle': {
							'color': '#333'
						}
					},
					'axisLabel': {
						'show': true,
						'textStyle': {
							'color': '#999999'
						}
					},
					'splitLine': {
						'show': true,
						'lineStyle': {
							'color': [
								'#eeeeee'
							]
						}
					},
					'splitArea': {
						'show': false,
						'areaStyle': {
							'color': [
								'rgba(250,250,250,0.05)',
								'rgba(200,200,200,0.02)'
							]
						}
					}
				},
				'logAxis': {
					'axisLine': {
						'show': true,
						'lineStyle': {
							'color': '#cccccc'
						}
					},
					'axisTick': {
						'show': false,
						'lineStyle': {
							'color': '#333'
						}
					},
					'axisLabel': {
						'show': true,
						'textStyle': {
							'color': '#999999'
						}
					},
					'splitLine': {
						'show': true,
						'lineStyle': {
							'color': [
								'#eeeeee'
							]
						}
					},
					'splitArea': {
						'show': false,
						'areaStyle': {
							'color': [
								'rgba(250,250,250,0.05)',
								'rgba(200,200,200,0.02)'
							]
						}
					}
				},
				'timeAxis': {
					'axisLine': {
						'show': true,
						'lineStyle': {
							'color': '#cccccc'
						}
					},
					'axisTick': {
						'show': false,
						'lineStyle': {
							'color': '#333'
						}
					},
					'axisLabel': {
						'show': true,
						'textStyle': {
							'color': '#999999'
						}
					},
					'splitLine': {
						'show': true,
						'lineStyle': {
							'color': [
								'#eeeeee'
							]
						}
					},
					'splitArea': {
						'show': false,
						'areaStyle': {
							'color': [
								'rgba(250,250,250,0.05)',
								'rgba(200,200,200,0.02)'
							]
						}
					}
				},
				'toolbox': {
					'iconStyle': {
						'normal': {
							'borderColor': '#999999'
						},
						'emphasis': {
							'borderColor': '#666666'
						}
					}
				},
				'legend': {
					'textStyle': {
						'color': '#999999'
					}
				},
				'tooltip': {
					'axisPointer': {
						'lineStyle': {
							'color': '#cccccc',
							'width': 1
						},
						'crossStyle': {
							'color': '#cccccc',
							'width': 1
						}
					}
				},
				'timeline': {
					'lineStyle': {
						'color': '#4ea397',
						'width': 1
					},
					'itemStyle': {
						'normal': {
							'color': '#4ea397',
							'borderWidth': 1
						},
						'emphasis': {
							'color': '#4ea397'
						}
					},
					'controlStyle': {
						'normal': {
							'color': '#4ea397',
							'borderColor': '#4ea397',
							'borderWidth': 0.5
						},
						'emphasis': {
							'color': '#4ea397',
							'borderColor': '#4ea397',
							'borderWidth': 0.5
						}
					},
					'checkpointStyle': {
						'color': '#4ea397',
						'borderColor': 'rgba(60,235,210,0.3)'
					},
					'label': {
						'normal': {
							'textStyle': {
								'color': '#4ea397'
							}
						},
						'emphasis': {
							'textStyle': {
								'color': '#4ea397'
							}
						}
					}
				},
				'visualMap': {
					'color': [
						'#d0648a',
						'#22c3aa',
						'#adfff1'
					]
				},
				'dataZoom': {
					'backgroundColor': 'rgba(255,255,255,0)',
					'dataBackgroundColor': 'rgba(222,222,222,1)',
					'fillerColor': 'rgba(114,230,212,0.25)',
					'handleColor': '#cccccc',
					'handleSize': '100%',
					'textStyle': {
						'color': '#999999'
					}
				},
				'markPoint': {
					'label': {
						'normal': {
							'textStyle': {
								'color': '#ffffff'
							}
						},
						'emphasis': {
							'textStyle': {
								'color': '#ffffff'
							}
						}
					}
				}
			});
	}

	onEChartsReady(evt: any) {
		console.log(evt);
	}

	onChartClick(evt: any) {
		console.log(evt);
	}
}
